rust/storage/src/caching.rs

```diff
@@ -0,0 +1,356 @@
+//! Our use of object storage is very specific in that we never overwrite objects and we think in
+//! blocks.  This module relies on this pattern to implement an object store that wraps two other
+//! object stores, one for the cache and one for the backing store.  Data is written to the backing
+//! store and then to the cache.  Reads are attempted from the cache and then from the backing
+//! store.  There is no effort made to dedupe writes, because it's assumed that a single writer is
+//! working on a single block at a time.  No one should write the same block to the same location
+//! at the same time, and the cost of doing so is cache inefficiency.
+
+use std::fmt::{Debug, Display};
+use std::hash::Hash;
+use std::ops::Range;
+use std::sync::Arc;
+
+use object_store::path::Path;
+use object_store::{
+    GetOptions, GetResult, GetResultPayload, ListResult, MultipartUpload, ObjectMeta, ObjectStore,
+    PutMultipartOpts, PutOptions, PutPayload, PutResult, Result,
+};
+use sync42::state_hash_table::{Key, StateHashTable, Value};
+
+use bytes::Bytes;
+use futures::stream::BoxStream;
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct DedupeKey {
+    path: Path,
+}
+
+impl Key for DedupeKey {}
+
+#[derive(Debug, Default)]
+pub struct DedupeValue {
+    one_in_flight: tokio::sync::Mutex<()>,
```

Comment:
that's clever! it gives you per key locking granularity easily without banging your head against the rust typing system. I'll use this pattern for spann!